runtime/v2/runner/src/main/java/com/walmartlabs/concord/runtime/v2/runner/DefaultEventReportingService.java

```diff
@@ -0,0 +1,154 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class DefaultEventReportingService implements EventReportingService {
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultEventReportingService.class);
+
+    private final InstanceId instanceId;
+    private final ProcessEventsApi processEventsApi;
+    private final Queue<ProcessEventRequest> eventQueue;
+    private final int maxBatchSize;
+    private final Object batchLock = new Object();
+    private final ScheduledExecutorService flushScheduler;
+
+    @Inject
+    public DefaultEventReportingService(InstanceId instanceId,
+                                        ProcessConfiguration processConfiguration,
+                                        ApiClient apiClient) {
+        this.instanceId = instanceId;
+        this.processEventsApi = new ProcessEventsApi(apiClient);
+        this.maxBatchSize = processConfiguration.events().batchSize();
+        this.eventQueue = new ArrayDeque<>(maxBatchSize);
+        this.flushScheduler = Executors.newSingleThreadScheduledExecutor();
+
+        int period = processConfiguration.events().batchFlushInterval();
+        flushScheduler.scheduleAtFixedRate(new FlushTimer(this), period, period, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void report(ProcessEventRequest req) {
+        synchronized (batchLock) {
+            eventQueue.add(req);
+        }
+
+        if (eventQueue.size() >= maxBatchSize) {
+            flush();
+        }
+    }
+
+    @Override
+    public void afterProcessEnds(Runtime runtime, State state, Frame lastFrame) {
+        flushScheduler.shutdown();
+        flush();
+    }
+
+    ProcessEventsApi getProcessEventsApi() {
+        return processEventsApi;
+    }
+
+    void flush() {
+        List<ProcessEventRequest> eventBatch = takeBatch();
+
+        while (!eventBatch.isEmpty()) {
+            if (eventBatch.size() == 1) {
+                sendSingle(eventBatch.get(0));
+            } else {
+                sendBatch(eventBatch);
+            }
+
+            eventBatch = takeBatch();
+        }
+    }
+
+    private void sendBatch(List<ProcessEventRequest> eventBatch) {
+        try {
+            getProcessEventsApi().batchEvent(instanceId.getValue(), eventBatch);
+        } catch (ApiException e) {
+            log.warn("Error while sending batch of {} event{} to the server: {}",
+                    eventBatch.size(), eventBatch.isEmpty() ? "" : "s", e.getMessage());
+        }
+    }
+
+    private void sendSingle(ProcessEventRequest req) {
+        try {
+            getProcessEventsApi().event(instanceId.getValue(), req);
+        } catch (ApiException e) {
+            log.warn("error while sending an event to the server: {}", e.getMessage());
+        }
+    }
+
+    /**
+     * @return batch of up-to {@link #maxBatchSize} queued process events
+     */
+    private List<ProcessEventRequest> takeBatch() {
```

Comment:
I think this might be an alternative.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/BlockingQueue.html#drainTo(java.util.Collection,int)

Reply:
I'm doing some thinking about that. The only downside is we have to be really strict about not over-filling the queue since that results in an exception with, say `ArrayBlockingQueue`--which may possibly happen with my first stab here when `parallel` blocks are running. It still reports up-to exactly the max batch size, but sometimes chatty threads can fill it up with more than that many events. Perhaps that's worth reigning in anyway.

Reply:
...or just use a `LinkedBlockingQueue` without a bound. Though I think I sorted out the possible routes of overfilling now.